go/mcap/indexed_message_iterator.go

```diff
@@ -162,34 +168,33 @@ func (it *indexedMessageIterator) loadChunk(chunkIndex *ChunkIndex) error {
 	case CompressionNone:
 		chunkData = parsedChunk.Records
 	case CompressionZSTD:
-		var err error
 		if it.zstdDecoder == nil {
-			it.zstdDecoder, err = zstd.NewReader(bytes.NewReader(parsedChunk.Records))
-		} else {
-			err = it.zstdDecoder.Reset(bytes.NewReader(parsedChunk.Records))
-		}
-		if err != nil {
-			return fmt.Errorf("failed to read zstd chunk: %w", err)
+			it.zstdDecoder, err = zstd.NewReader(nil)
+			if err != nil {
+				return fmt.Errorf("failed to instantiate zstd decoder: %w", err)
+			}
 		}
-		chunkData, err = io.ReadAll(it.zstdDecoder)
+		chunkData = make([]byte, 0, parsedChunk.UncompressedSize)
+		chunkData, err = it.zstdDecoder.DecodeAll(parsedChunk.Records, chunkData)
 		if err != nil {
-			return fmt.Errorf("failed to decompress zstd chunk: %w", err)
+			return fmt.Errorf("failed to decode chunk data: %w", err)
 		}
 	case CompressionLZ4:
 		if it.lz4Reader == nil {
 			it.lz4Reader = lz4.NewReader(bytes.NewReader(parsedChunk.Records))
 		} else {
 			it.lz4Reader.Reset(bytes.NewReader(parsedChunk.Records))
 		}
-		chunkData, err = io.ReadAll(it.lz4Reader)
+		chunkData = make([]byte, parsedChunk.UncompressedSize)
+		_, err = io.ReadFull(it.lz4Reader, chunkData)
 		if err != nil {
 			return fmt.Errorf("failed to decompress lz4 chunk: %w", err)
 		}
 	default:
 		return fmt.Errorf("unsupported compression %s", parsedChunk.Compression)
 	}
 	// use the message index to find the messages we want from the chunk
-	messageIndexSection := chunk[chunkIndex.ChunkLength:]
+	messageIndexSection := it.compressedChunk[chunkIndex.ChunkLength:compressedChunkLength]
```

Comment:
the naming is very confusing here - maybe we should change "compressedChunkLength" to "compressedChunkPlusIndexSectionLength".

Reply:
also clearly "compressedChunk" is not the compressed chunk record, it's the chunk plus the after-chunk data